### PR TITLE
storage: don't consider incomplete stores as replication targets

### LINKED ISSUE #5844: server: "NodeID must not be zero" errors after startup
build sha: multiple, up to and including 636e1b9

scenario: 5-node beta cluster running with different versions over 3 days.
when restarting nodes, I often see "NodeID must not be zero" errors for a few minutes after startup (not on all nodes though).

eg:
```
E160404 15:31:44.193962 storage/queue.go:446  [replicate] range=576 [/System/"tsd\x12cr.node.exec.error-10m\x00\x01\x89\xf8\x06/\xc62"-/System/"tsd\x12cr.node.exec.error-1h\x00\x01\x89\xf8\x06/\xb61"): error: storage/replica_command.go:2211: change replicas of 576 failed: roachpb/metadata.go:136: replica 3 is invalid: roachpb/metadata.go:153: NodeID must not be zero
```

I'm guessing some partially-filled request?

Full log for the latest restart of `ec2-54-209-133-121.compute-1.amazonaws.com`:
[node0.log.txt](https://github.com/cockroachdb/cockroach/files/202767/node0.log.txt)

### PR DESCRIPTION
**Pull request by bdarnell**
The store pool may contain references to incomplete stores that are not
yet considered dead, but these stores should not actually be used by the
replica allocator.

Fixes #5844

storage: make storeDetail.desc nullable
This gives incomplete storeDetails a clearly-invalid nil pointer instead
of a partial protobuf which can be used accidentally (as in #5844), and
removes the need for a separate boolean.

---

This change is [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/5856)

### CODE REVIEW

**Comment by petermattis** [comment]
:lgtm:

---

Review status: 0 of 3 files reviewed at latest revision, 3 unresolved discussions.

---

*[storage/store_pool.go, line 225 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5856#-KEZ4bdsVONcN0Vmu6Jl:-KEZ4bdsVONcN0Vmu6Jm:-120975556) ([raw file](https://github.com/cockroachdb/cockroach/blob/ab7cce6a550eb06dd7f56645bfc0dfab9fbf0d82/storage/store_pool.go#L225)):*
`s/GetStoreDescriptor/getStoreDetail/g`

---

*[storage/store_pool.go, line 228 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5856#-KEZ5clbAZ5xLVQSTKi1:-KEZ5clbAZ5xLVQSTKi2:1896087452) ([raw file](https://github.com/cockroachdb/cockroach/blob/ab7cce6a550eb06dd7f56645bfc0dfab9fbf0d82/storage/store_pool.go#L228)):*
Perusing this code now, seems like the locking should be moved up to `deadReplicas` and this method renamed to `getStoreDetailLocked`.

---

*[storage/store_pool.go, line 232 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5856#-KEZ5VMd3H7o8cN4U838:-KEZ5VMd3H7o8cN4U839:111089265) ([raw file](https://github.com/cockroachdb/cockroach/blob/ab7cce6a550eb06dd7f56645bfc0dfab9fbf0d82/storage/store_pool.go#L232)):*
Can you expand this comment and note when this situation can happen (e.g. on node startup before we've received gossip)?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/5856#-:-KEZ5qmTsrkZHyPm-eCE:1417088721)*

File: storage/store_pool.go [hunk 1]
```diff
@@ -325,11 +317,10 @@ func (sp *StorePool) getStoreList(required roachpb.Attributes, deterministic boo
 	var aliveStoreCount int
 	for _, storeID := range storeIDs {
 		detail := sp.stores[roachpb.StoreID(storeID)]
-		if !detail.dead {
+		if !detail.dead && detail.desc != nil {
```
Comment (BramGruneir) [inline storage/store_pool.go]: So adding `&& detail.gossiped` here would have achieved the same result?

**Comment by bdarnell** [comment]
Review status: 0 of 3 files reviewed at latest revision, 4 unresolved discussions.

---

*[storage/store_pool.go, line 225 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5856#-KEZ4bdsVONcN0Vmu6Jl:-KEbVIYig7npSb68phVS:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/ab7cce6a550eb06dd7f56645bfc0dfab9fbf0d82/storage/store_pool.go#L225)):*
Done.

---

*[storage/store_pool.go, line 228 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5856#-KEZ5clbAZ5xLVQSTKi1:-KEbVCkQwkhhnLcWCKNc:-711477219) ([raw file](https://github.com/cockroachdb/cockroach/blob/ab7cce6a550eb06dd7f56645bfc0dfab9fbf0d82/storage/store_pool.go#L228)):*
Done, although it's a bit subtle that this `get` method requires an `RWMutex` to be held in write mode. Explained in a comment.

---

*[storage/store_pool.go, line 232 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5856#-KEZ5VMd3H7o8cN4U838:-KEbVJJXZ0jITuQcl0aR:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/ab7cce6a550eb06dd7f56645bfc0dfab9fbf0d82/storage/store_pool.go#L232)):*
Done.

---

*[storage/store_pool.go, line 320 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5856#-KEYHjXQDslnPyH3W0ov-r1-320:-KEbR5UYGDNJe3IVknj7:1864543410) ([raw file](https://github.com/cockroachdb/cockroach/blob/ab7cce6a550eb06dd7f56645bfc0dfab9fbf0d82/storage/store_pool.go#L320)):*
Yes, and that's what I did in the first commit. The first commit fixes the bug; the second refactors and cleans up.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/5856)*

**Comment by BramGruneir** [comment]
:lgtm:

---

Review status: 0 of 3 files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/5856#-:-KEbY88LIHOdCAmZXRR2:1417088721)*

**Pull request merged**